### PR TITLE
Move to operator sdk version 1.24.0

### PR DESCRIPTION
**Pull request by ShyamsundarR**
Including update to gomega for golang.org/x/net
dependency update

### CODE REVIEW

**Comment by ShyamsundarR** [comment]
Operator SDK related version upgrade change as per documentation [here](https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.11.0/) are made in series.

Updated gomega in mod, which indirectly upgrades the dependent x/net to 0.7.0 (and other indirect dependents).

Changes to test are due to better error reporting, which as a result failed the tests that were updated.

There is one pending test failure, where a VRG is left behind and hence the API server does not stop, which will be fixed in a subsequent commit.

**Comment by nirs** [comment]
This is pretty big single commit pr. Can we split the work to smaller commits or multiple prs
to make this change incrementally?

**Comment by ShyamsundarR** [comment]
> This is pretty big single commit pr. Can we split the work to smaller commits or multiple prs to make this change incrementally?

Posted multiple commits...

**Comment by ShyamsundarR** [comment]
> Operator SDK related version upgrade change as per documentation [here](https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.11.0/) are made in series.
> 
> Updated gomega in mod, which indirectly upgrades the dependent x/net to 0.7.0 (and other indirect dependents).
> 
> Changes to test are due to better error reporting, which as a result failed the tests that were updated.
> 
> There is one pending test failure, where a VRG is left behind and hence the API server does not stop, which will be fixed in a subsequent commit.

Most of the above are now in commit messages, but note that `make test` still has the above failure and KUBEBUILDER_ASSETS is not set right from the Makefile on a new environment (i.e when run manually on the shell and then executing `make test` works, but on a fresh system it is empty)

**Comment by nirs** [comment]
Much nicer! The first commits updating the version looks trivial and
hopefully can be merge now, unless they are the reason for the test
failure.

Regarding gomega - if I understand this correctly, we update it for
fixing the security issue in /x/net. This looks fragile, for example
if we stop using this package, another package can bring older /x/net.

If we want to ensure that we have safe version of /x/net, why not make
the requirement explicit?

> Changes to test are due to better error reporting, which as a result failed the tests that were updated.
> There is one pending test failure, where a VRG is left behind and hence the API server does not stop, which will be fixed in a subsequent commit.

I don't understand the problem - the test failure are result of upgaraing
the versions or incomplete change, using new features available in the
new verson?

I think it would be useful to merge what we can merge now from this PR,
deferring the incomplete changes to later PR if possible.

**Comment by ShyamsundarR** [comment]
> Much nicer! The first commits updating the version looks trivial and hopefully can be merge now, unless they are the reason for the test failure.

I would like to merge this as a batch and not in parts. Having moved the SDK version from somewhere around 1.11 till here and picking changes that are intermediate, even though bisect-able, is preferred that we merge this in one go.
 
> 
> Regarding gomega - if I understand this correctly, we update it for fixing the security issue in /x/net. This looks fragile, for example if we stop using this package, another package can bring older /x/net.

Yes it can, `go mod graph` helps track down dependencies and address the same. Further security scans help check and report any new regressions.

> 
> If we want to ensure that we have safe version of /x/net, why not make the requirement explicit?

We do not use this package directly, it comes in as a dependency as an indirect package, so adding it directly does not serve the purpose, for similar reasons as above (i.e if we stop using say gomega, there would be no reason to add this to the module, although quite a few packages that we depend on also depend on /x/net).

> 
> > Changes to test are due to better error reporting, which as a result failed the tests that were updated.
> > There is one pending test failure, where a VRG is left behind and hence the API server does not stop, which will be fixed in a subsequent commit.
> 
> I don't understand the problem - the test failure are result of upgaraing the versions or incomplete change, using new features available in the new verson?

Upgrade of gomega now ends up returning status errors and compares them more correctly, causing test failures, requiring changes to this [file](https://github.com/RamenDR/ramen/pull/729/files#diff-8abc0d1caebc1310e1f171f866bc8d73a79a0e37fe451a3cea0d2ee3d435b5fd)

If you run the change in the commit without those changes the error reported by the tests are visible.

> 
> I think it would be useful to merge what we can merge now from this PR, deferring the incomplete changes to later PR if possible.

The changes are incomplete only from the perspective of why KUBEBUILDER_ASSETS is an empty string when run on a fresh setup, the "other" issue due to an existing issue in the VRG tests that was not considered a failure till now, but is from here on, which hence needs a fix.

I will fix these subsequently and update this PR for a merge, not spending time splitting this up.

**Comment by ShyamsundarR** [comment]
> > Operator SDK related version upgrade change as per documentation [here](https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.11.0/) are made in series.
> > Updated gomega in mod, which indirectly upgrades the dependent x/net to 0.7.0 (and other indirect dependents).
> > Changes to test are due to better error reporting, which as a result failed the tests that were updated.
> > There is one pending test failure, where a VRG is left behind and hence the API server does not stop, which will be fixed in a subsequent commit.
> 
> Most of the above are now in commit messages, but note that `make test` still has the above failure and [KUBEBUILDER_ASSETS](https://github.com/RamenDR/ramen/actions/runs/4257893544/jobs/7408523607#step:4:61) is not set right from the Makefile on a new environment (i.e when run manually on the shell and then executing `make test` works, but on a fresh system it is empty)

KUBEBUILDER_ASSETS issue in Makefile is fixed by splitting up the Makefile targets to envtest and setup-envtest, unsure why this works though.

envtests failure to shutdown the instance is addressed based on `https://github.com/kubernetes-sigs/controller-runtime/issues/1571`

All changes are updated and ready for review/merge as appropriate.

**Review by BenamarMk — APPROVED** [review]

File: controllers/drpolicy_controller_test.go [hunk 1]
```diff
@@ -289,12 +288,12 @@ var _ = Describe("DrpolicyController", func() {
...
 				)
-			}()
+			}
 			drp := drpolicy.DeepCopy()
 			drp.Spec.SchedulingInterval = `3s`
-			Expect(k8sClient.Create(context.TODO(), drp)).To(MatchError(err))
+			Expect(k8sClient.Create(context.TODO(), drp)).To(MatchError(err(drp.Spec.SchedulingInterval)))
 			drp.Spec.SchedulingInterval = `0`
-			Expect(k8sClient.Create(context.TODO(), drp)).To(MatchError(err))
+			Expect(k8sClient.Create(context.TODO(), drp)).To(MatchError(err(drp.Spec.SchedulingInterval)))
```
Comment (nirs) [inline controllers/drpolicy_controller_test.go:296]: Is this related to the envtest change?

File: controllers/drpolicy_controller_test.go [hunk 2]
```diff
@@ -313,15 +311,9 @@ var _ = Describe("DrpolicyController", func() {
...
+						field.Required(
 							path,
-							value,
-							validationErrors.InvalidType(
-								path.String(),
-								"body",
-								"array",
-								value,
-							).Error(),
+							"",
```
Comment (nirs) [inline controllers/drpolicy_controller_test.go:316]: Is this related to the envtest change?

File: config/crd/bases/ramendr.openshift.io_drclusters.yaml [hunk 3]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.9.0
+    controller-gen.kubebuilder.io/version: v0.9.2
```
[the remark this answers is not in the document]
Comment (ShyamsundarR) [reply config/crd/bases/ramendr.openshift.io_drclusters.yaml:6]: Looks like I did 2 things in one commit, this updates controller-gen version as per [this](https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.23.0/#gov3-update-controller-tools-from-090-to-092), causing the CRD changes that are auto generated when running make generate.

File: controllers/drpolicy_controller_test.go (hunk 1 above)
Comment (ShyamsundarR) [reply controllers/drpolicy_controller_test.go:296]: Yes, as we get a later envtest these fail when run in the latest environment. (same for the other comment)

File: Makefile [hunk 4]
```diff
@@ -338,7 +338,7 @@ ifeq (,$(shell which opm 2>/dev/null))
 	set -e ;\
 	mkdir -p $(dir $(OPM)) ;\
 	OS=$(shell go env GOOS) && ARCH=$(shell go env GOARCH) && \
-	curl -sSLo $(OPM) https://github.com/operator-framework/operator-registry/releases/download/v1.15.1/$${OS}-$${ARCH}-opm ;\
+	curl -sSLo $(OPM) https://github.com/operator-framework/operator-registry/releases/download/v1.23.2/$${OS}-$${ARCH}-opm ;\
```
[the remark this answers is not in the document]
Comment (ShyamsundarR) [reply Makefile:344]: True, this and other version strings within various tool downloads/use in the Makefile.

File: Makefile [hunk 5]
```diff
@@ -136,11 +136,14 @@ lint: golangci-bin ## Run configured golangci-lint and pre-commit.sh linters aga
...
-setup-envtest:
-	mkdir -p ${ENVTEST_ASSETS_DIR}
-	test -f ${ENVTEST_ASSETS_DIR}/setup-envtest.sh || curl -sSLo ${ENVTEST_ASSETS_DIR}/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/v0.8.3/hack/setup-envtest.sh
-	source ${ENVTEST_ASSETS_DIR}/setup-envtest.sh; fetch_envtest_tools $(ENVTEST_ASSETS_DIR); setup_envtest_env $(ENVTEST_ASSETS_DIR)
+envtest:
+	mkdir -p $(ENVTEST_ASSETS_DIR)
+	test -s $(ENVTEST_ASSETS_DIR)/setup-envtest || GOBIN=$(ENVTEST_ASSETS_DIR) go install sigs.k8s.io/controller-runtime/tools/setup-envtest@latest
+
+setup-envtest: envtest
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST_ASSETS_DIR)/setup-envtest use $(ENVTEST_K8S_VERSION) --bin-dir $(ENVTEST_ASSETS_DIR) -p path)"
```
[the remark this answers is not in the document]
Comment (ShyamsundarR) [reply Makefile:146]: The changed line is as per the new envtest in use and the suggested pattern by operator-sdk scaffolding.

When combined with the above as a single make target it does not work, when split as a separate target it does, which I have no explanation for.

File: Makefile [hunk 6]
```diff
@@ -38,7 +38,7 @@ IMAGE_NAME ?= ramen
 IMAGE_TAG ?= latest
 PLATFORM ?= k8s
 IMAGE_TAG_BASE = $(IMAGE_REGISTRY)/$(IMAGE_REPOSITORY)/$(IMAGE_NAME)
-RBAC_PROXY_IMG ?= "gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0"
+RBAC_PROXY_IMG ?= "gcr.io/kubebuilder/kube-rbac-proxy:v0.13.0"
```
[the remark this answers is not in the document]
Comment (ShyamsundarR) [reply Makefile:41]: This setting and changing things in the kustomize is to keep changed to committed sources to a minimum when deploying (which is where this image is used). IOW, this allows someone to use a different image, and when deployed would change that is in the deployment manifest via kustomize, if not changed then those manifests remain as is and not report any diffs when doing `git diff` etc.

Or, I am missing the question

File: config/crd/bases/ramendr.openshift.io_drplacementcontrols.yaml [hunk 7]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.9.0
+    controller-gen.kubebuilder.io/version: v0.9.2
```
[the remark this answers is not in the document]
Comment (nirs) [reply config/crd/bases/ramendr.openshift.io_drplacementcontrols.yaml:6]: It the files are auto generated, maybe they should not be part of git, and generated
when running make? I don't suggest to do this right now.

File: config/crd/bases/ramendr.openshift.io_drclusters.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (nirs) [reply config/crd/bases/ramendr.openshift.io_drclusters.yaml:6]: Ok, would be nice to mention this in the commit message.

File: Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (nirs) [reply Makefile:41]: So if you change this in the makefile, some autogenerated files are updated automatically
or you need to do the same change in multiple files (like we see in this commit)?

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (nirs) [reply Makefile:344]: I think we can do this later, maybe open an isue for this? this looks like a good first change.

File: config/manager/manager.yaml [hunk 8]
```diff
@@ -21,6 +21,8 @@ spec:
     metadata:
       labels:
         control-plane: controller-manager
+      annotations:
+        kubectl.kubernetes.io/default-container: manager
```
[the remark this answers is not in the document]
Comment (nirs) [reply config/manager/manager.yaml:25]: Sure it makes sense, just not explained in the commit message.

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (nirs) [reply Makefile:146]: I guess we will have to live with this for now.

**Comment by nirs** [comment]
> @nirs some of the changes are easier to follow if you run through a sample operator build exercise using [operator-sdk](https://sdk.operatorframework.io/docs/building-operators/golang/quickstart/). What I typically do is have a fresh sample created from the version we are moving to, to look at the changes/auto-generated files etc. and pull back changes to our code as per the upgrade documentation.
> 
> For example the envtest KUBEBUILDER_ASSETTS or the CRD annotations are explained away more easily.

Sounds like we need a maintainer.md document explaning this process for the next upgrade.

**Comment by ShyamsundarR** [comment]
@nirs Can we merge this now?

**Review by nirs — APPROVED** [review]

**Comment by nirs** [comment]
Sure, lets merge.
